### ISSUE TITLE
having 2 same index while drawing line chart

### ISSUE THREAD

**Issue by vb123er951** [facebookresearch/visdom#744]
**Bug Description**
Sometimes there have 2 same index with different values while drawing line chart
![image](https://user-images.githubusercontent.com/15091938/86425915-fa094e00-bd18-11ea-9096-0e66926bc615.png)

**Reproduction Steps**
use `vis.line()`, sometimes it works well but sometimes it will happen this kind of error.

**Expected behavior**
Every index has only 1 value.
![image](https://user-images.githubusercontent.com/15091938/86425886-e65de780-bd18-11ea-858e-86ec7ab36fe0.png)

**Server logs:**
something like these:
> INFO:tornado.access:200 POST /win_exists (127.0.0.1) 0.97ms
> INFO:tornado.access:200 POST /update (127.0.0.1) 1.00ms
> INFO:tornado.access:200 POST /win_exists (127.0.0.1) 1.00ms
> INFO:tornado.access:200 POST /update (127.0.0.1) 2.96ms
> INFO:tornado.access:200 POST /win_exists (127.0.0.1) 0.00ms
> INFO:tornado.access:200 POST /update (127.0.0.1) 1.02ms

**Additional context**
I am using pytorch with visdom, trying to print out the loss of model during training. I install visdom from source.

**Comment by vb123er951** [comment]
any comments?

**Comment by JackUrb** [comment]
Hi @vb123er951 - you haven't really provided enough context for me to know what's going on here. The screenshots of spreadsheets don't tell me how you're plotting, what your server setup is, or even what the plots are currently coming out as. The server logs look normal, so it's not that.

**Comment by vb123er951** [comment]
Hi @JackUrb ,
I use a class doing visdom drawing:

    class Visualization():
        def __init__(self, env_name='main'):
            self.vis = Visdom(
                port = 8097,
                server = 'http://localhost',
                base_url = '/',
                username = '',
                password = '',
                use_incoming_socket = True,
                env = env_name,
            )
            self.loss_window = None
            self.acc_window = None
            self.valloss_window = None
            self.valacc_window = None
            self.window_width = 600
            self.window_height = 300
            self.text_window = self.vis.text('Time process<br>start: {}'.format(time.strftime('%H:%M:%S')))
    
        def plot_(self, loss, acc1, acc5, step):
            y = np.column_stack((acc1, acc5))
            x = np.column_stack((step, step))
            self.acc_window = self.vis.line(
                Y = y, X = x,
                win = self.acc_window,
                update = 'append' if self.acc_window else None,
                opts = dict(
                    width = self.window_width,
                    height = self.window_height,
                    title = 'Training Accuracy',
                    xlabel = 'step',
                    ylabel = 'accuracy(%)',
                    legend = ['top 1 accuracy', 'top 5 accuracy'],
                    showlegend = True,
                )
            )
            self.loss_window = self.vis.line(
                Y = [loss],
                X = [step],
                win = self.loss_window,
                update = 'append' if self.loss_window else None,
                opts = dict(
                    width = self.window_width,
                    height = self.window_height,
                    title = 'Training Loss',
                    xlabel = 'step',
                    ylabel = 'loss',
                    linecolor = np.array([[50, 50, 255]])
                )
            )
    
        def plot_val(self, loss, acc1, acc5, step):
            y = np.column_stack((acc1, acc5))
            x = np.column_stack((step, step))
            self.valacc_window = self.vis.line(
                Y = y, X = x,
                win = self.valacc_window,
                update = 'append' if self.valacc_window else None,
                opts = dict(
                    width = self.window_width,
                    height = self.window_height,
                    title = 'Validation Accuracy',
                    xlabel = 'step',
                    ylabel = 'accuracy(%)',
                    legend = ['top 1 accuracy', 'top 5 accuracy'],
                    showlegend = True,
                )
            )
            self.valloss_window = self.vis.line(
                Y = [loss],
                X = [step],
                win = self.valloss_window,
                update = 'append' if self.valloss_window else None,
                opts = dict(
                    width = self.window_width,
                    height = self.window_height,
                    title = 'Validation Loss',
                    xlabel = 'step',
                    ylabel = 'loss',
                    # linecolor = np.array([[50, 50, 255]])
                )
            )
    
        def plot_text(self, content):
            self.vis.text('{}'.format(content), win=self.text_window, append=True)

and during training, I create a instance of this class:
`vis = Visualization('tasn')`

and in the training loop, use pytorch model to predict output and calculate loss:

    l = loss.item()
    vis.plot_(l, top1, top5, step)
    step += 1
top1, top5 are the top1 accuracy and top5 accuracy calculated by pytorch function

sometimes the plot shows:
![newplot](https://user-images.githubusercontent.com/15091938/98621776-f9299e00-2342-11eb-97b4-2bf12a601aa9.png)
it suppose to be as:
![trainloss](https://user-images.githubusercontent.com/15091938/98621887-40b02a00-2343-11eb-97db-e6b078943abc.png)
(the train loss and validation loss are only for example, they are not same numbers)

**Comment by JackUrb** [comment]
Hi @vb123er951 - this may be due to a bug in how visdom assigns window ids - you can resolve this by specifying the window ID even during the first creation. For example:
```
self.loss_window = self.vis.line(
            Y = [loss],
            X = [step],
            win = f'loss_window_experiment_{exp}',
            update = 'append',
            opts = dict(
                width = self.window_width,
                height = self.window_height,
                title = 'Training Loss',
                xlabel = 'step',
                ylabel = 'loss',
                linecolor = np.array([[50, 50, 255]])
            )
        )
```
You also no longer need to specify update to `append` or `None` if you'll always be appending, append on a non-existent window is now treated as a create.

**Comment by vb123er951** [comment]
Hi @JackUrb,
thank you for your advise, I'll try it!

**Comment by Manik-Khajuria-5** [comment]
It has been resolved visdom now raises an error instead of silently creating duplicate entries
